### PR TITLE
[FIX] sale_coupon_advanced: remove discount lines with initial line

### PR DESCRIPTION
**Pull request by oerp-odoo**
Improving the way discount lines are handled, when related promotion order line is removed. It will not remove related discount line, if promotion line is removed (before, you would need to remove it manually).

### CODE REVIEW

File: sale_coupon_advanced/models/sale_order_line.py
Comment (i-vyshnevska) [inline sale_coupon_advanced/models/sale_order_line.py]:
```suggestion
            programs = SaleCouponProgram.search([
```
there can be a lot of them

File: sale_coupon_advanced/models/sale_order_line.py
Comment (i-vyshnevska) [inline sale_coupon_advanced/models/sale_order_line.py:21]:
```suggestion
        for line in self.filtered(lambda self: self.forced_reward_line):
```

File: sale_coupon_advanced/models/sale_order_line.py
Comment (i-vyshnevska) [inline sale_coupon_advanced/models/sale_order_line.py]: maybe try to call super method first and then remove just discount line

**Review by i-vyshnevska — COMMENTED** [review]
can you apply pre-commit, I want to check with which errors travis fails

File: sale_coupon_advanced/models/sale_order_line.py
Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_order_line.py]: Right, but then I need to modify logic too. I somehow thought there would be only one. I need to be aware of that when filtering.

File: sale_coupon_advanced/models/sale_order_line.py
Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_order_line.py:21]: As it is now, it should take care of all rewards, not just forced ones. Or we only want to remove related discount for forced lines?

File: sale_coupon_advanced/models/sale_order_line.py
Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_order_line.py]: Actually tried it, but getting same error.

**Comment by StephaneMangin** [comment]
Can you indicate the card reference plz

**Comment by oerp-odoo** [comment]
I was told to not indicate references on public PRs:)

File: sale_coupon_advanced/models/sale_order.py
Comment (StephaneMangin) [inline sale_coupon_advanced/models/sale_order.py:118]: I'm kind of perplex about the complexity of this behavior... Is it not a bit complicated to 'just' remove a discount line ?

Comment (StephaneMangin) [reply sale_coupon_advanced/models/sale_order.py:118]: Does an unlink over the discount line not suffice ?

When i read that "But because we add it automatically (without needing to add that product, because we force it), it nullifies this standard behavior." : we would expect to 'fix' directly the behavior instead of adding a new level of code ? no ?

File: sale_coupon_advanced/models/sale_order_line.py
Comment (StephaneMangin) [reply sale_coupon_advanced/models/sale_order_line.py:21]: If the problem comes from our module, it should only be applied on the behavior it's made for. So @i-vyshnevska is right.

File: sale_coupon_advanced/models/sale_order.py
Comment (StephaneMangin) [reply sale_coupon_advanced/models/sale_order.py:118]: Have you tried to remove the program associated to the discount line that remains after removing the product and apply the update procedure of the normal behavior ?

**Comment by ajaniszewska-dev** [comment]
> I was told to not indicate references on public PRs:)

then it is ok to write 2-3 sentences in PR description and explain what is the aim/scope of your PR, otherwise not always clear what you wanted to do.

File: sale_coupon_advanced/models/sale_order_line.py
Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_order_line.py:21]: Hm, its not that its the problem, it is functionality. I can only apply to to forced lines, but then it will gimp functionality. Is there a reason, user would want to keep non forced discount line, if related promotion line was removed? If not, then I see no reason to use it only for forced line, cause it could be useful in both cases.

File: sale_coupon_advanced/models/sale_order.py
Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_order.py:118]: Unfortunately its not that simple. First iteration implementation was done just like that. But there are side effects that happen when you try to unlink lines via Odoo interface (Form). It tries to use same line that is being deleted (by onchange spec). So we decided to make workaround that we clean up write if such things happens. You can check tests, specifically the one with Form. Thats special case, where all this complexity originates from.

Comment (oerp-odoo) [reply sale_coupon_advanced/models/sale_order.py:118]: @StephaneMangin if by normal procedure you mean `recompute coupon lines`, then it wont work. As you already found my comment, it works the opposite of what we need. It adds missing lines, instead of removing. So in that case, we won't know if lines need to be added or removed.

**Review by StephaneMangin — APPROVED** [review]
Seen with @chroi @p-tombez and @yvaucher. We can't let this PR to block the release process. Can you just add a succinct explanation of the issue and add a FIXME to your write method?

**Comment by oerp-odoo** [comment]
Sure, will do.

**Comment by oerp-odoo** [comment]
@StephaneMangin still need one more approval. Or you already discussed this and I can proceed with merge?

**Review by chroi — APPROVED** [review]
lg

**Pull request merged**
